---
layout: workshop_main_5day
permalink: /BiCG_2019
title: Bioinformatics for Cancer Genomics 2019
header1: Workshop Pages for Students
header2: Bioinformatics for Cancer Genomics 2019
image: /site_images/CBW_cancerDNA_icon-16.jpg
keywords: Visualize genomic data, pathway analysis, integrate clinical data
description: Course covers the key bioinformatics concepts and tools required to analyze cancer genomic data sets and access and work with data sets in the cloud. 
instructors: Jared Simpson, Trevor Pugh, Francis Ouellette, Sorana Morrissy, Florence Cavalli, Juri Reimand, Lincoln Stein, Robin Haw, Mark Phillips, Zhibin Lu
length: 5 days
---

# Welcome <a id="welcome"></a>

Welcome to Bioinformatics for Cancer Genomics 2019.  

The course schedule can be found [here](https://bioinformaticsdotca.github.io/BiCG_2019_schedule).

If you have questions or comments, visit the workshop [Q&A forum](https://classroom.google.com). You may need to join the group by clicking on the "+" symbol and entering the class code.

**CLASS MATERIALS:** You can download the printed course manual [here]().  

***

# Pre-Workshop Materials <a id="preworkshop"></a>

## Laptop Setup Instructions

Instructions to setup your laptop can be found [here](https://bioinformaticsdotca.github.io/laptop_setup_instructions).

## Pre-workshop Tutorials

Pre-readings and pre-work can be found here [here](https://bioinformaticsdotca.github.io/BiCG_2019_prework).
It is in your best interest to complete these before the workshop.

1) **R Preparation tutorials**: You are expected to have completed the following tutorials in **R** beforehand. The tutorial should be very accessible even if you have never used **R** before.

* The [CBW R tutorial](http://bioinformatics-ca.github.io/CBW_R_Tutorial/)
* [Quick and Dirty Guide to **R**](http://ww2.coastal.edu/kingw/statistics/R-tutorials/text/quick&dirty_R.txt)  
* The [R Tutorial](http://www.cyclismo.org/tutorial/R/) up to and including 5. Basic Plots
* The [R command cheat sheet](../../resources/R_Short-refcard.pdf)

2) **UNIX Preparation tutorials**:  

* [UNIX Bootcamp](https://rnabio.org/module-00-setup/0000/06/01/Unix/)
* Tutorials #1-3 on [UNIX Tutorial for Beginners](http://www.ee.surrey.ac.uk/Teaching/Unix/)  
* [Unix Cheat sheet](http://www.rain.org/~mkummel/unix.html) 

3) [Sequencing Terminology](http://www.ncbi.nlm.nih.gov/projects/genome/glossary.shtml)

4)  **Cytoscape Preparation tutorials**: Complete the [introductory tutorial to Cytoscape](http://opentutorials.cgl.ucsf.edu/index.php/Portal:Cytoscape3)

* Introduction to Cytoscape3 - User Interface

* Introduction to Cytoscape3 - Welcome Screen

* Filtering and Editing in Cytoscape 3

## Pre-workshop Readings  

Before coming to the workshop, read these:
  
  * [Integrative Genomics Viewer (IGV): high-performance genomics data visualization and exploration](http://www.ncbi.nlm.nih.gov/pubmed/22517427)
  
  * [Genome structural variation discovery and genotyping](http://www.ncbi.nlm.nih.gov/pubmed/21358748)
  
  * [A survey of sequence alignment algorithms for next-generation sequencing](http://www.ncbi.nlm.nih.gov/pubmed/20460430)
  
  * [Genotype and SNP calling from next-generation sequencing data](http://www.ncbi.nlm.nih.gov/pubmed/21587300)
  
  * [Informatics for RNA Sequencing: A Web Resource for Analysis on the Cloud](https://www.ncbi.nlm.nih.gov/pubmed/26248053)  

  * [Transcript-level expression analysis of RNA-seq experiments with HISAT, StringTie and Ballgown](https://www.ncbi.nlm.nih.gov/pubmed/27560171)
  
  * [ENCODE RNA-Seq Standards](https://genome.ucsc.edu/ENCODE/protocols/dataStandards/ENCODE_RNAseq_Standards_V1.0.pdf)
  
  * [Methods to study splicing from high-throughput RNA sequencing data](http://www.ncbi.nlm.nih.gov/pubmed/24549677)  
  
  * [A comprehensive assessment of RNA-seq accuracy, reproducibility and information content by the Sequencing Quality Control Consortium](http://www.ncbi.nlm.nih.gov/pubmed/25150838)

  * [De novo transcript sequence reconstruction from RNA-seq using the Trinity platform for reference generation and analysis](http://www.nature.com/nprot/journal/v8/n8/full/nprot.2013.084.html)
  
  * [Recurrent chimeric RNAs enriched in human prostate cancer identified by deep sequencing](http://www.ncbi.nlm.nih.gov/pubmed/21571633)
  
  * [Pathway enrichment analysis and visualization of omics data using g:Profiler, GSEA, Cytoscape and EnrichmentMap](https://www.nature.com/articles/s41596-018-0103-9)
  
  * [iRegulon: From a Gene List to a Gene Regulatory Network Using Large Motif and Track Collections](http://www.ncbi.nlm.nih.gov/pubmed/25058159)
  
  * [The GeneMANIA prediction server: biological network integration for gene prioritization and predicting gene function](http://www.ncbi.nlm.nih.gov/pubmed/20576703)
  
  * [GeneMANIA Prediction Server 2013 Update](http://www.ncbi.nlm.nih.gov/pubmed/23794635)
  
  * [How to visually interpret biological data using networks](http://www.ncbi.nlm.nih.gov/pubmed/19816451)
  
  * [g:Profiler--a web-based toolset for functional profiling of gene lists from large-scale experiments](http://www.ncbi.nlm.nih.gov/pubmed/17478515)
  
  * [g:Profiler--a web server for functional interpretation of gene lists (2011 update)](http://www.ncbi.nlm.nih.gov/pubmed/21646343)
  
  * [Gene set enrichment analysis: a knowledge-based approach for interpreting genome-wide expression profiles](http://www.ncbi.nlm.nih.gov/pubmed/16199517)
  
  * [Expression data analysis with reactome](http://www.ncbi.nlm.nih.gov/pubmed/25754994)  

***

# Class Photo

Coming soon  

*** 

# Day 1 <a id="day1"></a>

## Welcome

*<font color="#827e9c">Michelle Brazas</font>*

## Module 1: Introduction to Cancer Genomics

*<font color="#827e9c">Trevor Pugh</font>* 

[Lecture]()  

## Module 2: Ethics of Data Usage and Security

*<font color="#827e9c">Mark Phillips</font>*

[Lecture]()

## Module 3: Databases and Visualization Tools

*<font color="#827e9c">Francis Ouellette</font>* 

[Lecture]() 

## Module 3 Lab: IGV

*<font color="#827e9c">Heather Gibling</font>*

[Lab Practical]() 

## Logging into the Cloud

*<font color="#827e9c">Zhibin Lu</font>*

[Lecture]()  

[Lab Practical: Connecting to the Cloud](http://bioinformaticsdotca.github.io/AWS_setup)

* We have set up instances on the Amazon cloud - one for each student. In order to log in to your instance, you will need a security certificate. If you plan on using Linux or Mac OS X, please download [this certificate](http://main.oicrcbw.ca/private/CBWNY.pem). Otherwise if you plan on using Windows (with Putty), please download [this certificate](http://main.oicrcbw.ca/private/CBWNY.ppk).  

## Unix and R review

*<font color="#827e9c">Heather Gibling</font>*

[Assignment](https://bioinformaticsdotca.github.io/bicg-2018-Unix_and_R)  

* [Data file 1](https://raw.githubusercontent.com/bioinformaticsdotca/Genomic_Med_2017/master/test.fasta)  
* [Data file 2](https://raw.githubusercontent.com/bioinformaticsdotca/BiCG_2017/master/Review_session/Gene_R_example.csv)  

***

# Day 2 <a id="day2"></a>

## Module 4: Genome Alignment

*<font color="#827e9c">Jared Simpson</font>* 

[Lecture]()  

[Lab Practical]()  

## Module 5: Genome Assembly

*<font color="#827e9c">Jared Simpson</font>* 

[Lecture]()

[Lab Practical]()  

## Module 6: Copy Number Variants

*<font color="#827e9c">Sorana Morrissy</font>* 

[Lecture]()  

## Integrative Assignment  

[Assignment](https://bioinformaticsdotca.github.io/bicg_2018_ia2)

***

# Day 3 <a id="day3"></a>  

## Module 6 Lab

*<font color="#827e9c">Sorana Morrissy</font>* 

[Lab Lecture]()  

[Lab Practical]()  

* [Installation](https://bioinformaticsdotca.github.io/bicg_2018_module6_lab_install)  
* [Data Preprocessing](https://bioinformaticsdotca.github.io/bicg_2018_module6_lab_preprocessing)

## Module 7: Somatic Mutations and Annotations

*<font color="#827e9c">Sorana Morrissy</font>* 

[Lecture](https://drive.google.com/a/bioinformatics.ca/file/d/1loSEwod3wt-gs5xJF-RcgbeUHdSUls-U/view?usp=sharing)  
  
[Lab Practical](https://bioinformaticsdotca.github.io/bicg_2018_lab_7)  

## Module 8: Gene Expression Profiling 

*<font color="#827e9c">Florence Cavalli</font>* 

[Lecture]()  

## Integrated Assignment

[Integrated Assignment]()  

***

# Day 4 <a id="day4"></a>

## Module 8 Lab  

*<font color="#827e9c">Florence Cavalli</font>* 

[Ballgown code](https://drive.google.com/a/bioinformatics.ca/file/d/0B2tkpG9cB9bdbFlCeHEzcDBDSmZ4aTZ4dHpiQUVET0c1TU1j/view?usp=sharing)  

[Supplementary code](https://drive.google.com/a/bioinformatics.ca/file/d/0B2tkpG9cB9bdZndNNy1LYVVWNEJaT0x3Ykp3WTQteW9jTzhv/view?usp=sharing)

[Supplementary R output](https://drive.google.com/a/bioinformatics.ca/file/d/0B2tkpG9cB9bdNWZ2UEtUSDlTaVM2VGpWMTFQMkRoamFLYXdF/view?usp=sharing)

[Questions with Answers PDF](https://drive.google.com/a/bioinformatics.ca/file/d/1qvrULzdtZ8SIU28kLMl4E3M_XFo94QyM/view?usp=sharing)  

[Questions with Answers Word](https://drive.google.com/a/bioinformatics.ca/file/d/1Zstc8oAqLgppQF-wdvJzHZU6xbYBNAjc/view?usp=sharing)  

## Module 9: Gene Fusion and Rearrangements

*<font color="#827e9c">Brian Haas</font>* 

[Lecture](https://drive.google.com/a/bioinformatics.ca/file/d/1YM4fh1CZNZHmUeITCfvw8y1-dKZPHxOj/view?usp=sharing)  

[Recording](https://youtu.be/TzjxFkDHO4M)

[Lab Practical - CTAT Fusion Toolkit](https://bioinformaticsdotca.github.io/BiCG_2018_module9_star_fusion)

[Lab Practical - DISCASM and GMAP-Fusion](https://bioinformaticsdotca.github.io/BiCG_2018_module9_discasm)

## Module 10: Genes to Pathways

*<font color="#827e9c">JÃ¼ri Reimand</font>* 

[Lecture]()  

[Lab Practical - gProfiler](https://bioinformaticsdotca.github.io/bicg_2018_lab_8_gprofiler)  

## Integrated Assignment

[Integrated Assignment]()  

***

# Day 5 <a id="day5"></a>  

## Module 11: Variants to Networks

*<font color="#827e9c">Robin Haw</font>* 

[Lecture]()

[Lab Lecture]()

[Lab Practical]()  

## Module 12: Big Cancer Genomics

*<font color="#827e9c">Lauren Erdman</font>* 

[Lecture]()

[Lab Practical](https://drive.google.com/a/bioinformatics.ca/file/d/1FZy349vf-e3G3MDeowRjNYDJJ-AwjeX7/view?usp=sharing)  

[Lab Script](https://drive.google.com/a/bioinformatics.ca/file/d/1Wxz3BRWvu2iAFhJB-voz44mGSAZ0vI9l/view?usp=sharing)  

[Data for lab part 1](https://github.com/bioinformaticsdotca/BiCG_2017/raw/master/module9/CBW-CancerGenomics-June2017-Lab-Data-Half1.RData)   
[Data for lab part 2](https://github.com/bioinformaticsdotca/BiCG_2017/raw/master/module9/CBW-CancerGenomics-June2017-Lab-Data-Half2.RData)   

## Recap

[Overview of workshop](https://bioinformaticsdotca.github.io/bicg_2018_recap)  

***

Thank you for attending the Bioinformatics for Cancer Genomics workshop! Help us make this workshop better by filling out [our survey]().

